wire-runtime/src/commonMain/kotlin/com/squareup/wire/MessageSource.kt

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire
+
+expect interface MessageSource<out T> {
+  fun receiveOrNull(): T?
+
+  fun close()
+}
+
+expect inline fun <T> MessageSource<T>.consumeEachAndClose(block: (T) -> Unit)
```

Comment:
Hmm I'm actually tempted to leave this as an internal function of the client for now. I don't think we expose anything this DSL-like for anything else, do we?